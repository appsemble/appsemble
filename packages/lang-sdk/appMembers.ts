export function assignAppMemberProperties(properties: any, formData: FormData): void {
  if (properties && typeof properties === 'object' && !Array.isArray(properties)) {
    formData.append(
      'properties',
      JSON.stringify(
        Object.fromEntries(
          Object.entries(properties).map(([key, value]) => [
            key,
            typeof value === 'string' ? value : JSON.stringify(value),
          ]),
        ),
      ),
    );
  }
}
