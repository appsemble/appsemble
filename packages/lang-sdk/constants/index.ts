export * from './baseTheme.js';
export * from './fonts.js';
export * from './locale.js';
export * from './patterns.js';
