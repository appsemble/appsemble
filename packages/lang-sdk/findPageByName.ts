import { type PageDefinition } from './types/index.js';

export function findPageByName(pages: PageDefinition[], name: string): PageDefinition | undefined {
  for (const page of pages) {
    if (page.name === name) {
      return page;
    }

    if (page.type === 'container' && page.pages) {
      const found = findPageByName(page.pages, name);
      if (found) {
        return found;
      }
    }
  }

  return undefined;
}
