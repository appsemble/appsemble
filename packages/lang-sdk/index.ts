export * as schemas from './schemas/index.js';
export * from './remap.js';
export * from './miscellaneous.js';
export * from './constants/index.js';
export * from './normalize.js';
export * from './authorization.js';
export * from './blockUtils.js';
export * from './validation.js';
export * from './findPageByName.js';
export * from './appMessages.js';
export * from './iterApp.js';
export * from './mapValues.js';
export * from './validateAppMessages.js';
export * from './serverActions.js';
export * from './allActions.js';
export * from './appMembers.js';
export * from './examples.js';
export * from './noop.js';
export * from './string.js';
export * from './jsonschema.js';
export * as referenceSchemas from './reference-schemas/index.js';
