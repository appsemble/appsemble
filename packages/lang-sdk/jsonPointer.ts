// From https://gitlab.com/remcohaszing/koas/-/blob/main/packages/koas-core/src/jsonRefs.ts

/**
 * Escape a JSON pointer segment.
 *
 * See https://tools.ietf.org/html/rfc6901#section-3
 *
 * @param pointer THe JSON pointer segment to escape.
 * @returns The escaped JSON pointer segment.
 */
export function escapeJsonPointer(pointer: string): string {
  return pointer.replaceAll('~', '~0').replaceAll('/', '~1');
}

/**
 * Unescape a JSON pointer segment.
 *
 * See https://tools.ietf.org/html/rfc6901#section-3
 *
 * @param pointer The JSON pointer segment to unescape
 * @returns The unescaped JSON pointer segment.
 */
export function unescapeJsonPointer(pointer: string): string {
  return pointer.replaceAll('~1', '/').replaceAll('~0', '~');
}
