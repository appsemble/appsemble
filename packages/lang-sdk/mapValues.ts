export function mapValues<Input, Output>(
  object: Record<string, Input>,
  iteratee: (value: Input) => Output,
): Record<string, Output> {
  const result: Record<string, Output> = {};
  for (const [key, value] of Object.entries(object)) {
    result[key] = iteratee(value);
  }
  return result;
}
