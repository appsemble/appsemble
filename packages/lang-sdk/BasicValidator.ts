import { type Options, type Schema, Validator, type ValidatorResult } from 'jsonschema';

import { schemas as allSchemas } from './index.js';
import { escapeJsonPointer } from './jsonPointer.js';
import { has } from './miscellaneous.js';

interface ValidatorFactoryOptions {
  schemas: Record<string, Schema>;
  customFormats?: Record<string, (value: unknown) => boolean>;
}

export class BaseValidatorFactory {
  static readonly defaultOptions: Options = {
    base: '#',
    nestedErrors: true,
  };

  private schemas;

  private customFormats;

  constructor({ customFormats, schemas }: ValidatorFactoryOptions) {
    this.customFormats = customFormats;
    this.schemas = schemas;
  }

  build(): Validator {
    const validator = new Validator();

    for (const [key, format] of Object.entries(this.customFormats ?? {})) {
      validator.customFormats[key] = format;
    }

    for (const [key, schema] of Object.entries(this.schemas)) {
      const path = `#/components/schemas/${escapeJsonPointer(key)}`;
      validator.addSchema(schema, path);
    }

    return validator;
  }
}

export class BlockExampleValidator {
  private validator: Validator;

  constructor() {
    this.validator = new BaseValidatorFactory({
      schemas: allSchemas,
      customFormats: {
        fontawesome: () => true,
        remapper: () => true,
        action: () => true,
        'event-listener': () => true,
        'event-emitter': () => true,
      },
    }).build();
  }

  validate(example: any): ValidatorResult {
    const { required, ...blockSchema } = structuredClone(allSchemas!.BlockDefinition) as Schema;

    delete blockSchema.properties?.name;
    delete blockSchema.properties?.version;

    const actionsSchema = blockSchema.properties!.actions as Schema;

    delete actionsSchema?.additionalProperties;
    if (example.actions) {
      // TODO: do we handle `$any` appropriately here?
      actionsSchema.properties = Object.fromEntries(
        Object.keys(example.actions).map((key) => [
          key,
          { $ref: '#/components/schemas/ActionDefinition' },
        ]),
      );
    }
    const blockEventsSchema = {
      type: 'object',
      additionalProperties: false,
      properties: {} as {
        emit?: Schema;
        listen?: Schema;
      },
    } satisfies Schema;
    blockSchema.properties!.events = blockEventsSchema;
    if (example.events) {
      if (example.events.emit) {
        blockEventsSchema.properties.emit = has(example.events.emit, '$any')
          ? { type: 'object', additionalProperties: { type: 'string' } }
          : {
              type: 'object',
              properties: Object.fromEntries(
                Object.keys(example.events.emit).map((emitter) => [emitter, { type: 'string' }]),
              ),
            };
      }
      if (example.events.listen) {
        blockEventsSchema.properties.listen = has(example.events.listen, '$any')
          ? { type: 'object', additionalProperties: { type: 'string' } }
          : {
              type: 'object',
              properties: Object.fromEntries(
                Object.keys(example.events.listen).map((listener) => [
                  listener,
                  { type: 'string' },
                ]),
              ),
            };
      }
    }

    // TODO: do we validate example parameters properly?

    const validationResult = this.validator.validate(
      example,
      blockSchema,
      BaseValidatorFactory.defaultOptions,
    );
    return validationResult;
  }
}

export class BlockParamValidator {
  private validator: Validator;

  constructor() {
    this.validator = new BaseValidatorFactory({
      schemas: allSchemas,
      customFormats: {
        fontawesome: () => true,
        remapper: () => true,
        action: () => true,
        'event-listener': () => true,
        'event-emitter': () => true,
      },
    }).build();
  }

  validateParametersSchema(paramSchema: unknown): ValidatorResult {
    return this.validator.validate(
      paramSchema,
      allSchemas.JSONSchemaRoot,
      BaseValidatorFactory.defaultOptions,
    );
  }
}

export class AppValidator {
  private validator: Validator;

  constructor() {
    this.validator = new BaseValidatorFactory({
      schemas: allSchemas,
    }).build();
  }

  validateApp(app: unknown): ValidatorResult {
    return this.validator.validate(
      app,
      allSchemas.AppDefinition,
      BaseValidatorFactory.defaultOptions,
    );
  }
}
