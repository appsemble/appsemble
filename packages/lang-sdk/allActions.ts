import { schemas } from '@appsemble/lang-sdk';
import { type ActionDefinition } from '@appsemble/types';
import { type OpenAPIV3 } from 'openapi-types';

// Import { schemas } from './index.js';

type NonArraySchemaObject = OpenAPIV3.NonArraySchemaObject;
type ReferenceObject = OpenAPIV3.ReferenceObject;
type SchemaObject = OpenAPIV3.SchemaObject;

const [, kinds] = schemas.ActionDefinition.allOf as NonArraySchemaObject[];

export const allActions = new Set(
  (kinds.anyOf as ReferenceObject[])
    .map(({ $ref }) => {
      const ref = $ref.split('/').at(-1);
      const action = schemas[ref as keyof typeof schemas] as SchemaObject;
      const name = (action.properties?.type as SchemaObject).enum?.[0] as string | undefined;
      return name;
    })
    .filter((x) => x !== undefined),
);

export type ActionName = ActionDefinition['type'];
