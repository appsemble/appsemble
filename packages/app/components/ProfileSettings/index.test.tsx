import { type AppDefinition } from '@appsemble/lang-sdk';
import { render } from '@testing-library/react';
import { IntlProvider } from 'react-intl';
import { beforeEach, describe, expect, it, vi } from 'vitest';

import { ProfileSettings } from './index.js';
import * as appDefinitionProvider from '../AppDefinitionProvider/index.js';
import * as appMemberProvider from '../AppMemberProvider/index.js';

const appMemberInfo = {
  sub: 'randomly-generated',
  email_verified: true,
  demo: false,
  role: 'User',
  zoneinfo: 'Europe/Amsterdam',
  email: 'test@appsemble.com',
  name: 'Test User',
  $seed: false,
  $ephemeral: false,
};

const appDefinition: AppDefinition = {
  name: 'Test App',
  layout: { enabledSettings: ['name', 'picture', 'languages'] },
  defaultPage: 'Test Page',
  pages: [],
};

const navigate = vi.fn();
vi.mock('react-router-dom', async () => ({
  ...(await vi.importActual('react-router-dom')),
  useNavigate: () => navigate,
}));

beforeEach(() => {
  vi.clearAllMocks();
});

describe('ProfileDropdown', () => {
  it('should render fields mentioned in the app definition', () => {
    vi.spyOn(appMemberProvider, 'useAppMember').mockImplementation(() => ({
      logout: vi.fn(),
      passwordLogin: vi.fn(),
      authorizationCodeLogin: vi.fn(),
      demoLogin: vi.fn(),
      isLoggedIn: true,
      role: 'User',
      appMemberGroups: [{ id: 1, name: 'Test Group', role: 'User' }],
      appMemberInfo,
      // @ts-expect-error null is not assignable to MutableRefObject
      appMemberInfoRef: null,
      addAppMemberGroup: vi.fn(),
      setAppMemberInfo: vi.fn(),
      setAppMemberSelectedGroup: vi.fn(),
      appMemberSelectedGroup: { id: 1, name: 'Test Group', role: 'User' },
    }));

    vi.spyOn(appDefinitionProvider, 'useAppDefinition').mockImplementation(() => ({
      definition: appDefinition,
      demoMode: false,
      revision: 1,
      blockManifests: [],
    }));

    const { container } = render(
      <IntlProvider locale="en" messages={{}}>
        <ProfileSettings />
      </IntlProvider>,
    );
    expect(container.getElementsByClassName('field')).toMatchSnapshot();
  });

  it('should render phoneNumber field', () => {
    vi.spyOn(appMemberProvider, 'useAppMember').mockImplementation(() => ({
      logout: vi.fn(),
      passwordLogin: vi.fn(),
      authorizationCodeLogin: vi.fn(),
      demoLogin: vi.fn(),
      isLoggedIn: true,
      role: 'User',
      appMemberGroups: [{ id: 1, name: 'Test Group', role: 'User' }],
      appMemberInfo,
      // @ts-expect-error null is not assignable to MutableRefObject
      appMemberInfoRef: null,
      addAppMemberGroup: vi.fn(),
      setAppMemberInfo: vi.fn(),
      setAppMemberSelectedGroup: vi.fn(),
      appMemberSelectedGroup: { id: 1, name: 'Test Group', role: 'User' },
    }));

    Object.assign(appDefinition, {
      members: {
        phoneNumber: {
          enable: true,
        },
      },
      layout: { enabledSettings: ['name', 'picture', 'languages', 'phoneNumber'] },
    });
    vi.spyOn(appDefinitionProvider, 'useAppDefinition').mockImplementation(() => ({
      definition: appDefinition,
      demoMode: false,
      revision: 1,
      blockManifests: [],
    }));

    const { container } = render(
      <IntlProvider locale="en" messages={{}}>
        <ProfileSettings />
      </IntlProvider>,
    );
    expect(container.getElementsByClassName('field')).toMatchSnapshot();
  });

  it('should render no fields', () => {
    vi.spyOn(appMemberProvider, 'useAppMember').mockImplementation(() => ({
      logout: vi.fn(),
      passwordLogin: vi.fn(),
      authorizationCodeLogin: vi.fn(),
      demoLogin: vi.fn(),
      isLoggedIn: true,
      role: 'User',
      appMemberGroups: [{ id: 1, name: 'Test Group', role: 'User' }],
      appMemberInfo,
      // @ts-expect-error null is not assignable to MutableRefObject
      appMemberInfoRef: null,
      addAppMemberGroup: vi.fn(),
      setAppMemberInfo: vi.fn(),
      setAppMemberSelectedGroup: vi.fn(),
      appMemberSelectedGroup: { id: 1, name: 'Test Group', role: 'User' },
    }));

    Object.assign(appDefinition, {
      layout: { enabledSettings: [] },
    });
    vi.spyOn(appDefinitionProvider, 'useAppDefinition').mockImplementation(() => ({
      definition: appDefinition,
      demoMode: false,
      revision: 1,
      blockManifests: [],
    }));

    const { container } = render(
      <IntlProvider locale="en" messages={{}}>
        <ProfileSettings />
      </IntlProvider>,
    );
    expect(container.getElementsByClassName('field')).toMatchInlineSnapshot('HTMLCollection []');
  });
});
