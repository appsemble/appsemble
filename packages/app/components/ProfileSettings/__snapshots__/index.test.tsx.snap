// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProfileDropdown > should render fields mentioned in the app definition 1`] = `
HTMLCollection [
  <div
    class="field"
  >
    <label
      class="label"
      for="name"
    >
      Display Name
      <span
        class="is-pulled-right has-text-weight-normal"
      >
        (
        Optional
        )
      </span>
    </label>
    <div
      class="control control has-icons-left"
    >
      <input
        class="input"
        id="name"
        name="name"
        placeholder="Display Name"
        value="Test User"
      />
      <span
        class="icon is-left"
      >
        <i
          class="fas fa-user"
        />
      </span>
    </div>
    <span
      class="help help"
    >
      This is the name that displays for other users.
    </span>
  </div>,
  <div
    class="field"
  >
    <label
      class="label"
      for="picture"
    >
      Picture
      <span
        class="is-pulled-right has-text-weight-normal"
      >
        (
        Optional
        )
      </span>
    </label>
    <div
      class="control control"
    >
      <figure
        class="image is-128x128 mb-2"
      >
        <img
          alt="Picture"
          class="picture is-rounded"
        />
      </figure>
      <div
        class="file has-name"
      >
        <label
          class="file-label undefined"
          for="picture"
        >
          <input
            accept="image/jpeg, image/png, image/tiff, image/webp"
            class="file-input"
            id="picture"
            name="picture"
            type="file"
          />
          <span
            class="file-cta"
          >
            <span
              class="icon"
            >
              <i
                class="fas fa-upload"
              />
            </span>
            <span
              class="file-label"
            >
              Picture
            </span>
          </span>
          <span
            class="file-name"
          >
            Select a file
          </span>
        </label>
      </div>
      <p
        class="help"
      >
        The public picture to use for your profile.
      </p>
    </div>
    <span
      class="help help"
    />
  </div>,
  <div
    class="field"
  >
    <div
      class="control control"
    >
      <div>
        <div>
          <button
            class="button mr-1"
            type="button"
          >
            Click a Picture
          </button>
          <button
            class="button"
            disabled=""
            type="button"
          >
            <span
              class="icon is-large"
            >
              <i
                class="fas fa-camera fa-2x"
              />
            </span>
          </button>
        </div>
        <p
          class="help"
        >
          Click a picture straight from your camera, (Only works if no file is being uploaded)
        </p>
      </div>
    </div>
    <span
      class="help help"
    />
  </div>,
  <div
    class="field"
  >
    <label
      class="label"
      for="locale"
    >
      Language
      <span
        class="is-pulled-right has-text-weight-normal"
      >
        (
        Optional
        )
      </span>
    </label>
    <div
      class="control control has-icons-left"
    >
      <div
        class="select is-fullwidth"
      >
        <select
          class="is-fullwidth"
          id="locale"
          name="locale"
          size="1"
        >
          <option
            value="en"
          >
            English (Supported)
          </option>
          <option
            value="nl"
          >
            Dutch (Nederlands)
          </option>
        </select>
      </div>
      <span
        class="icon is-left"
      >
        <i
          class="fas fa-language"
        />
      </span>
    </div>
    <span
      class="help help"
    >
      Preferred language for this app
    </span>
  </div>,
]
`;

exports[`ProfileDropdown > should render phoneNumber field 1`] = `
HTMLCollection [
  <div
    class="field"
  >
    <label
      class="label"
      for="name"
    >
      Display Name
      <span
        class="is-pulled-right has-text-weight-normal"
      >
        (
        Optional
        )
      </span>
    </label>
    <div
      class="control control has-icons-left"
    >
      <input
        class="input"
        id="name"
        name="name"
        placeholder="Display Name"
        value="Test User"
      />
      <span
        class="icon is-left"
      >
        <i
          class="fas fa-user"
        />
      </span>
    </div>
    <span
      class="help help"
    >
      This is the name that displays for other users.
    </span>
  </div>,
  <div
    class="field"
  >
    <label
      class="label"
      for="phoneNumber"
    >
      Phone Number
      <span
        class="is-pulled-right has-text-weight-normal"
      >
        (
        Optional
        )
      </span>
    </label>
    <div
      class="control control has-icons-left"
    >
      <input
        class="input"
        id="phoneNumber"
        name="phoneNumber"
        placeholder="Phone Number"
        value=""
      />
      <span
        class="icon is-left"
      >
        <i
          class="fas fa-phone"
        />
      </span>
    </div>
    <span
      class="help help"
    />
  </div>,
  <div
    class="field"
  >
    <label
      class="label"
      for="picture"
    >
      Picture
      <span
        class="is-pulled-right has-text-weight-normal"
      >
        (
        Optional
        )
      </span>
    </label>
    <div
      class="control control"
    >
      <figure
        class="image is-128x128 mb-2"
      >
        <img
          alt="Picture"
          class="picture is-rounded"
        />
      </figure>
      <div
        class="file has-name"
      >
        <label
          class="file-label undefined"
          for="picture"
        >
          <input
            accept="image/jpeg, image/png, image/tiff, image/webp"
            class="file-input"
            id="picture"
            name="picture"
            type="file"
          />
          <span
            class="file-cta"
          >
            <span
              class="icon"
            >
              <i
                class="fas fa-upload"
              />
            </span>
            <span
              class="file-label"
            >
              Picture
            </span>
          </span>
          <span
            class="file-name"
          >
            Select a file
          </span>
        </label>
      </div>
      <p
        class="help"
      >
        The public picture to use for your profile.
      </p>
    </div>
    <span
      class="help help"
    />
  </div>,
  <div
    class="field"
  >
    <div
      class="control control"
    >
      <div>
        <div>
          <button
            class="button mr-1"
            type="button"
          >
            Click a Picture
          </button>
          <button
            class="button"
            disabled=""
            type="button"
          >
            <span
              class="icon is-large"
            >
              <i
                class="fas fa-camera fa-2x"
              />
            </span>
          </button>
        </div>
        <p
          class="help"
        >
          Click a picture straight from your camera, (Only works if no file is being uploaded)
        </p>
      </div>
    </div>
    <span
      class="help help"
    />
  </div>,
  <div
    class="field"
  >
    <label
      class="label"
      for="locale"
    >
      Language
      <span
        class="is-pulled-right has-text-weight-normal"
      >
        (
        Optional
        )
      </span>
    </label>
    <div
      class="control control has-icons-left"
    >
      <div
        class="select is-fullwidth"
      >
        <select
          class="is-fullwidth"
          id="locale"
          name="locale"
          size="1"
        >
          <option
            value="en"
          >
            English (Supported)
          </option>
          <option
            value="nl"
          >
            Dutch (Nederlands)
          </option>
        </select>
      </div>
      <span
        class="icon is-left"
      >
        <i
          class="fas fa-language"
        />
      </span>
    </div>
    <span
      class="help help"
    >
      Preferred language for this app
    </span>
  </div>,
]
`;
